Library/ViewModels/CommentsViewModel.swift

```diff
@@ -220,17 +259,93 @@ public final class CommentsViewModel: CommentsViewModelType,
+  public let cellSeparatorHidden: Signal<Bool, Never>
   public let commentComposerViewHidden: Signal<Bool, Never>
   public let configureCommentComposerViewWithData: Signal<CommentComposerViewData, Never>
-  public let errorMessage: Signal<String, Never>
   public let goToCommentReplies: Signal<(Comment, Project), Never>
-  public let isCommentsLoading: Signal<Bool, Never>
-  public let isCellSeparatorHidden: Signal<Bool, Never>
   public let loadCommentsAndProjectIntoDataSource: Signal<([Comment], Project), Never>
-  public let postCommentSuccessful: Signal<Comment, Never>
-  public var postCommentSubmitted: Signal<(), Never>
-  public let shouldShowLoadingIndicator: Signal<Bool, Never>
+  public let resetCommentComposerAndScrollToTop: Signal<(), Never>
+  public let showLoadingIndicatorInFooterView: Signal<Bool, Never>
 
   public var inputs: CommentsViewModelInputs { return self }
   public var outputs: CommentsViewModelOutputs { return self }
 }
+
+private func retryCommentProducer(
+  erroredComment comment: Comment,
+  project: Project
+) -> SignalProducer<(Comment, String), Never> {
+  // Retry posting the comment.
+  AppEnvironment.current.apiService.postComment(
+    input: .init(
+      body: comment.body,
+      commentableId: project.graphID
+    )
+  )
+  .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+  // Return a producer with the successful comment that is prefixed by a comment indicating that the
+  // retry was successful and then, after a 1 second delay, the actual successful comment is returned.
+  .flatMap { successfulComment -> SignalProducer<Comment, ErrorEnvelope> in
+    let retrySuccessComment = successfulComment.updatingStatus(to: .retrySuccess)
+      |> \.id .~ comment.id // Inject the original errored comment's ID to replace.
+
+    return SignalProducer(value: successfulComment)
+      .ksr_delay(.seconds(1), on: AppEnvironment.current.scheduler)
+      .prefix(value: retrySuccessComment)
+  }
+  // Immediately return a comment in a retrying state when this producer starts.
+  .prefix(value: comment.updatingStatus(to: .retrying))
+  // If retrying errors again, return the original comment returning it to its errored state.
+  .demoteErrors(replaceErrorWith: comment)
+  // Return the comment that will be replaced with the ID to find it by.
+  .map { replacementComment in (replacementComment, comment.id) }
+}
+
+private func postCommentProducer(
+  project: Project,
+  user: User,
+  body: String
+) -> SignalProducer<(Comment, String), Never> {
+  let failableComment = Comment.failableComment(
+    withId: AppEnvironment.current.uuidType.init().uuidString,
+    date: AppEnvironment.current.dateType.init().date,
+    project: project,
+    user: user,
+    body: body
+  )
+
+  // Post the new comment.
+  return AppEnvironment.current.apiService.postComment(
+    input: .init(
+      body: body,
+      commentableId: project.graphID
```

Comment:
> What is `GraphIDBridging`?

Just a protocol that ensures that this type can construct a relay ID for matching in our schema. Relay IDs have the structure of `Model-<id>` so in this case like `Comment-12345`.

Reply:
> In general, I was just wondering how the `id` of the `Comment` is created before its' `failed` status is confirmed by the network request? Because my assumption was a `Comment` has no `id` until it is `successful`. Just my own knowledge gap there.

This refers to the ID of the project to associate the comment to it. The project is the `Commentable` here.